### PR TITLE
Add "inline svg" and "d3" plugins

### PR DESCRIPTION
**Pull request by vincerubinetti**
- adds plugin to "inline" svg img's (finds  tags with .svg src's, and replaces them in-place with the text source content of the svg)
- adds d3 plugin
- adds hash and other security attributes to mathjax plugin

Supercedes #373

A few notes:
- this wont work on local copies due to cross origin, but should fail gracefully, and not change the `` if it can't load the source text. to test this PR, you'll have to run a local server, either with [the manubot commands](https://github.com/manubot/rootstock#local-execution) or this handy [chrome app](https://chrome.google.com/webstore/detail/web-server-for-chrome/ofhbbkphhbklhfoeikjpcbhemlocgigb/related?hl=en) 
- i've split these into two separate plugins, because they're really two separate concerns
- inserting the d3 script with javascript shouldn't be necessary because pandoc provides a way to insert things into the document
- we can use cleaner async/await because it's supported in all the browsers Manubot supports
- changed some things to be more consistent with the other plugins (use of `let`/`const` instead of `var`, comments at top of file, etc)
- checking for `data` attribute is unnecessary because we're only querying for ``'s, not `<object>`'s
- getting `` whose `src` attribute ends in `.svg` can all be done as a CSS selector
- using d3's filter and each functions is unnecessary because javascript built-in funcs will do fine. this way the plugins can be separate, too
- from my testing and from numerous of online examples, the`d3.js` source `<script>` tag can be in the `<head>` or `<body>`, as long as it comes before the code that uses d3. thus, i'm including the d3 plugin before the body. 
- no "onload" event listener of the d3 library is needed. but in our case, we're also doing our own asynchronous processing, to inline the svgs, so any code that relies on inlined svgs will have to wait until a `D3Loaded` event is triggered
- in order to transfer over attributes from the original image to the new SVG, we must parse the svg with a `DOMParser` rather than use `outerHTML`. idk why this is, but i dont think you can set additional attributes of an element after setting its `outerHTML`. anyway, this has the added benefit of checking to see if it's valid svg syntaxt too, rather than blindly copy/pasting its source into the document

### CODE REVIEW

**Comment by vincerubinetti** [comment]
Apologies for the multiple pushes.

Upon testing, this works, but it messes up a few things. The lightbox plugin wasn't designed to work on inline SVGs (which may not be a problem). Also some of the styles in the theme seem to assume an `img` tag and not a `svg` tag. Finally, this makes it so that the width/height the user sets for the image is lost, but I think I can remedy that.

**Comment by vincerubinetti** [comment]
Okay, with the last 2 commits I was able to solve the latter 2 problems from my previous comment. This is ready for review @dhimmel @agitter @matthewturk

**Comment by matthewturk** [comment]
I checked for `data` attributes to try to reduce the chances of picking up, say, ORCID and whatnot, and trying to restrict them to just figure images.  If there were a way to specify additional discriminating attributes to some of the figures, but not necessarily all of them, that would be fine too.

Anyway, I'll defer to everyone else on this PR.

**Comment by vincerubinetti** [comment]
@dhimmel or @agitter I think this is ready to be merged. I tested it and it works, though I'd prefer if other people could test it. But we can merge it now and just update it if anyone finds issues. It's a fairly simple plugin.

**Comment by dhimmel** [comment]
Cool. Will take a look this weekend and try it out

**Review by dhimmel — COMMENTED** [review]
Looks good to me. @matthewturk your review is probably more meaningful here... does this address your use case?

@vincerubinetti Are there any behavior changes to the webpage in the default configuration (with d3 and inline-svg disabled)?

> If there were a way to specify additional discriminating attributes to some of the figures, but not necessarily all of them, that would be fine too.

@vincerubinetti any thoughts? @matthewturk do we need to do something before merging to address this?

ORCID logos is just an `` and not a `<figure>` right? Is that helpful?

Another option is to use the pandoc syntax for attributes, see [link_attributes](https://pandoc.org/MANUAL.html#extension-link_attributes) in the docs:

> Attributes can be set on links and images: `An inline image{#id .class width=30 height=20px}`

Thanks @matthewturk for #373. When merging, we'll note your contribution [with](https://docs.github.com/en/free-pro-team@latest/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors):

```
Co-authored-by: Matthew Turk <matthewturk@gmail.com>
```

File: build/plugins/d3.html
Comment (vincerubinetti) [reply build/plugins/d3.html:1]: It doesn't necessarily depend on the inline svg plugin. One could write a basic `<svg>` container right in the markdown, or create it with d3.

File: build/pandoc/defaults/html.yaml
Comment (vincerubinetti) [reply build/pandoc/defaults/html.yaml:6]: Idk why it would have an effect but we can comment it out sure.

**Comment by vincerubinetti** [comment]
> @vincerubinetti Are there any behavior changes to the webpage in the default configuration (with d3 and inline-svg disabled)?

Not that I know of.

> > If there were a way to specify additional discriminating attributes to some of the figures, but not necessarily all of them, that would be fine too.

We could make it such that you have to put a special attribute on the svg for it to be inlined. But given this is such a niche need already, I don't see that it would matter. There is no downside (nothing you lose) when you inline an SVG, you only gain functionality. The only downside is inconvenience of inlining.

Thus it doesn't matter if we inline the ORCID and twitter icons, if that's what Matt was talking about. I didn't understand his comment because my original comment was about not checking for the `data` attribute, which is something that applies more to `<object>`, not ``.

**Comment by vincerubinetti** [comment]
One small note though, when you inline an SVG, to select it with a CSS style, you'll have to do `svg` instead of `img`. But that should mostly only be relevant to us the developers.

**Review by agitter — COMMENTED** [review]
I don't have any strong opinions about this, so I'm okay with it if it works for @matthewturk's use case.

The CI build has an error we should resolve
```
Error parsing build/pandoc/defaults/html.yaml line 5 column 20:
expected !!str instead of !!null
Error: Process completed with exit code 64.
```
Possibly that's related to https://github.com/manubot/rootstock/pull/375/files#r505102115

**Comment by matthewturk** [comment]
This works for me, yup.

**Review by agitter — APPROVED** [review]

**Comment by dhimmel** [comment]
Thanks @vincerubinetti and @matthewturk. Looking forward to seeing these plugins in action.

**Pull request merged**
